### PR TITLE
8313693: Introduce an internal utility for the Damerau–Levenshtein distance calculation

### PR DESCRIPTION
**Pull request by pavelrappo**
Please review this PR to introduce the Damerau–Levenshtein distance calculation for later use in REPL and CLI.

The Damerau–Levenshtein distance (DL) is a string similarity metric, variants of which have been successfully used in REPL and CLI:

    $ git hlpe
    git: 'hlpe' is not a git command. See 'git --help'.
    
    The most similar commands are
	    grep
	    help

or

    $ hg clonmit
    hg: unknown command 'clonmit'
    (did you mean one of clone, commit, config?)

Originally, DL was requested by https://bugs.openjdk.org/browse/JDK-8288660 to provide helpful error messages for mistyped javadoc tags. However, given its wider applicability (e.g. mistyped command-line options), it was suggested that DL is put somewhere more internally accessible, for example, in `com.sun.tools.javac.util`, which is available in `jdk.compiler` and already exported to some other potential clients:

    exports com.sun.tools.javac.util to
            jdk.jdeps,
            jdk.javadoc,
            jdk.jshell;

---

The Levenshtein edit distance is concerned with three types of edits: insertions, deletions, and substitutions. The Damerau-Levenshtein distance adds the fourth type: transpositions of two adjacent characters. This considerably complicates the algorithm, which is currently known in two main forms: restricted and unrestricted.

The difference between the two forms is that the simpler form, restricted, operates under the assumption that no substring is edited more than once. This results in different distances between some strings, for example, "CA" and "ABC". Restricted DL gives 3, whereas unrestricted DL gives 2.

Git seems to have [chosen](https://git.kernel.org/pub/scm/git/git.git/commit/?id=8af84dadb142f7321ff0ce8690385e99da8ede2f) restricted DL. This PR proposes **un**restricted DL. From my nonexpert perspective, while unrestricted DL does not necessarily guarantee a better end result, it feels a better place to start. Additionally, unrestricted DL is a proper metric. That property could probably be used to quickly estimate the upper bound of DL in some applications.

Like restricted DL, unrestricted DL runs in `O(s1.length() * s2.length())` in time. However, unrestricted DL requires more space: `O(s1.length() * s2.length())` plus the space for the alphabet, which for the purposes stated in this PR should be relatively small: `[a-zA-Z0-9]` plus maybe a dozen of additional symbols, such as `_` and `-`.

The algorithm that this PR proposes is my port of the algorithm from a 1975 paper freely available at https://dl.acm.org/doi/10.1145/321879.321880, with a few Java-related tweaks which when benchmarked seem to speed up (x1.3) the ASCII strings case.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8313693](https://bugs.openjdk.org/browse/JDK-8313693): Introduce an internal utility for the Damerau–Levenshtein distance calculation (**Bug** - P4)

### Reviewers
 * [Jan Lahoda](https://openjdk.org/census#jlahoda) (@lahodaj - **Reviewer**)
 * [Jonathan Gibbons](https://openjdk.org/census#jjg) (@jonathan-gibbons - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15157/head:pull/15157` \
`$ git checkout pull/15157`

Update a local copy of the PR: \
`$ git checkout pull/15157` \
`$ git pull https://git.openjdk.org/jdk.git pull/15157/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15157`

View PR using the GUI difftool: \
`$ git pr show -t 15157`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15157.diff">https://git.openjdk.org/jdk/pull/15157.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15157#issuecomment-1665620525)

### CODE REVIEW

File: test/langtools/tools/javac/util/StringUtilsTest.java [hunk 1]
```diff
@@ -53,6 +55,57 @@ void run() throws Exception {
         assertEquals(2, StringUtils.indexOfIgnoreCase("  lookFor", "lookfor"));
         assertEquals(11, StringUtils.indexOfIgnoreCase("  lookFor  LOOKfor", "lookfor", 11));
         assertEquals(2, StringUtils.indexOfIgnoreCase("\u0130\u0130lookFor", "lookfor"));
+
+        //verify Damerau-Levenshtein
+
+        assertEquals(3, DamerauLevenshteinDistance.of("kitten", "sitting"));
+        // note that the unrestricted Damerau-Levenshtein distance would be 3, not 2:
```
Comment (archiecobbs) [inline test/langtools/tools/javac/util/StringUtilsTest.java:62]: I think the "unrestricted" in this comment should instead be "restricted".

**Comment by pavelrappo** [comment]
I was asked elsewhere to provide demonstration of intended use, for example in jdk.javadoc. Okay:
```
        var knownTags = """
                author
                code
                deprecated
                docRoot
                exception
                hidden
                index
                inheritDoc
                link
                linkplain
                literal
                param
                provides
                return
                see
                serial
                serialData
                serialField
                since
                snippet
                summary
                systemProperty
                throws
                uses
                value
                version"""
                .lines().toList();

        record Pair(String tag, int distance) { }

        for (var misspelled : """
                serial-filed
                inherotDoc
                inheritdoc
                param
                parm
                returns
                throw
                depreciated
                """.lines().toList()) {
            knownTags.stream()
                    .map(tag -> new Pair(tag, DamerauLevenshteinDistance.of(misspelled, tag)))
                    .filter(i -> Double.compare(1 / 3.0, ((double) i.distance()) / i.tag().length()) >= 0)
                    .sorted(Comparator.comparingDouble(Pair::distance))
                    .limit(3)
                    .forEach(p -> System.out.println(misspelled + "\n\tdid you mean " + p.tag() + "?"));
        }
```
results in this
```
serial-filed
	did you mean serialField?
inherotDoc
	did you mean inheritDoc?
inheritdoc
	did you mean inheritDoc?
param
	did you mean param?
parm
	did you mean param?
returns
	did you mean return?
throw
	did you mean throws?
depreciated
	did you mean deprecated?
```

File: test/langtools/tools/javac/util/StringUtilsTest.java (hunk 1 above)
Comment (pavelrappo) [reply test/langtools/tools/javac/util/StringUtilsTest.java:62]: Fixed; thanks!

**Comment by JimLaskey** [comment]
Brilliant application BTW.

File: test/langtools/tools/javac/util/StringUtilsTest.java [hunk 2]
```diff
@@ -66,4 +119,10 @@ void assertEquals(int expected, int actual) {
             throw new IllegalStateException("expected=" + expected + "; actual=" + actual);
         }
     }
+
+    void assertTrue(boolean cond) {
+        if (!cond) {
+            throw new IllegalStateException();
```
Comment (jonathan-gibbons) [inline test/langtools/tools/javac/util/StringUtilsTest.java:125]: Mildly surprising to see `IllegalStateException` in `assert...` methods.

Comment (pavelrappo) [reply test/langtools/tools/javac/util/StringUtilsTest.java:125]: I simply followed the style of the file.

**Comment by pavelrappo** [comment]
>         for (var misspelled : """
>                 serial-filed
>                 inherotDoc
>                 inheritdoc
>                 param

(snip)

> ```
> param
> 	did you mean param?
> ```

I was asked elsewhere about "param". It's a typo that should've been "params", which is a real-world misspelling captured in https://bugs.openjdk.org/browse/JDK-8288660.

**Comment by pavelrappo** [comment]
> (FWIW, I have some prototype for suggesting similar names for mistypes elements, although I was always on the fence on how useful is that in the real world.)

It remains to be seen how useful it will be. I'd imagine it'll be helpful to [on-ramp audience](https://openjdk.org/projects/amber/design-notes/on-ramp) in CLI and REPL. Additionally, it might be helpful for mistyped tags in jdk.javadoc. Unlike command-line options, the list of javadoc tags is not the easiest thing to obtain. Even if it were, sometimes even seasoned developers occasionally stumble upon `@returns` or `@throw`. In which case, a helpful error message will likely reduce head-scratching.

**Pull request closed without merging**